---
title: "STR variation analysis (primary tumors)"
author: "Max Verbiest"
date: '2022-03-21'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load libs}
library(tidyverse)
```

```{r load panels, echo=FALSE}
# load tral_perf panel for exonic regions
df_exon_panel <- read.delim("/Users/maxverbiest/PhD/data/str_panels/tral_and_perf_panel_exons.tsv", header=FALSE, na.strings=".") %>% 
  cbind(., str_split_fixed(.$V7, "[:]", n=Inf)) %>% 
  dplyr::select(-c(V5, V6, V7, `3`)) %>% 
  mutate(
    `1` = ifelse(`1` == ".", NA, `1`),
    `4` = as.integer(`4`)
    )
colnames(df_exon_panel) <- c("chr", "start", "end", "period", "repeat_id", "msa", "max_p_stretch")

df_exon_panel <- df_exon_panel %>% 
  mutate(
    tmp_id = paste0(chr, "_", start),
    ref = (end - start + 1) / period
  )

# load tral_perf panel for CDS regions
df_cds_panel <- read.delim("/Users/maxverbiest/PhD/data/str_panels/tral_and_perf_panel_CDS.tsv", header=FALSE, na.strings=".") %>% 
  cbind(., str_split_fixed(.$V7, "[:]", n=Inf)) %>% 
  dplyr::select(-c(V5, V6, V7, `3`)) %>% 
  mutate(
    `1` = ifelse(`1` == ".", NA, `1`),
    `4` = as.integer(`4`)
    )

colnames(df_cds_panel) <- c("chr", "start", "end", "period", "repeat_id", "msa", "max_p_stretch")
df_cds_panel <- df_cds_panel %>% 
  mutate(
    tmp_id = paste0(chr, "_", start),
    ref = (end - start + 1) / period
  )
```

```{r load variation, echo=FALSE}
# load variation data
df_variation <- read.csv("/Users/maxverbiest/PhD/projects/genotype_strs/results/oxana_variation/crc_variations_march.csv")
df_variation <- df_variation %>% 
  distinct() %>%  # Goes from 351157 -> 351064 after distinct(). Why?
  # mutate(alt = as.integer(alt), ref = as.integer(ref)) %>% 
  mutate(tmp_id = paste0(chr, "_", start), ref_len_diff = abs(alt - ref))

df_clinical <- read.csv("/Users/maxverbiest/PhD/projects/genotype_strs/data/oxana_tcga_clinical.csv", header = TRUE) %>%
  dplyr::select(c(case_submitter_id, project_id, gender, MSI, CMS))
colnames(df_clinical)[1] <- "patient"

# pivot into wider table where allele A and B are on the same row
df_patient_len_diff <- df_variation %>% 
  group_by(patient, tmp_id) %>%
  mutate(allele = LETTERS[1:n()]) %>%
  pivot_wider(
    id_cols = c(patient, sample_type, tmp_id),
    names_from = allele,
    values_from = alt,
    names_prefix = "allele_"
  ) %>% 
  mutate(allele_B = if_else(is.na(allele_B), allele_A, allele_B)) %>% 
  ungroup()

df_patient_len_diff <- df_patient_len_diff %>% 
  left_join(df_variation %>% select(tmp_id, period, ref) %>% distinct(), by="tmp_id") %>% 
  mutate(
    patient_len_diff = abs((ref - allele_A)) + abs((ref - allele_B))
    )

df_mutated_strs <- df_patient_len_diff %>% 
  dplyr::filter(patient_len_diff > 0)
```

```{r}
df_patient_lef_diff <- read.csv("/Users/maxverbiest/PhD/projects/genotype_strs/results/oxana_variation/crc_variations_march_wide.tsv", sep="\t", header=TRUE)
```


```{r}
df_summary <- df_patient_len_diff %>% 
  group_by(patient) %>% 
  summarize(
    calls = n()
  ) %>% 
  left_join(
    df_mutated_strs %>% 
      group_by(patient) %>% 
      summarize(
        nonref_calls = n(), 
        total_len_diff = sum(patient_len_diff), 
        mean_len_dif = total_len_diff/nonref_calls
        ), by="patient") %>% 
  left_join(
    df_clinical %>% dplyr::select(patient, MSI, CMS),
    by = "patient"
  ) %>% 
  mutate(frac_nonref = nonref_calls / calls)

write.table(df_summary, "../../results/oxana_variation/crc_variations_march_summary.tsv", row.names = FALSE, sep="\t", quote=FALSE)
```

```{r}
df_patient_len_diff <- df_patient_len_diff %>% 
  left_join(
    df_exon_panel %>% select(tmp_id) %>% distinct() %>% mutate(in_exon = TRUE),
    by="tmp_id"
  ) %>%
  mutate(in_exon = if_else(is.na(in_exon), FALSE, TRUE))

df_patient_len_diff <- df_patient_len_diff %>% 
  left_join(
    df_cds_panel %>% select(tmp_id) %>% distinct() %>% mutate(in_cds = TRUE),
    by="tmp_id"
  ) %>%
  mutate(in_cds = if_else(is.na(in_cds), FALSE, TRUE))
```

```{r}
df_variation <- df_variation %>% 
  left_join(
    df_cds_panel %>% select(tmp_id) %>% distinct() %>% mutate(in_cds = TRUE),
    by="tmp_id"
  ) %>%
  mutate(in_cds = if_else(is.na(in_cds), FALSE, TRUE))

df_variation %>% 
  dplyr::filter(ref_len_diff <= 10, ref_len_diff > 0) %>%
  ggplot(aes(x=alt-ref, fill=in_exon)) +
  geom_histogram(position = position_dodge(), stat="count") +
  facet_wrap(~period, scale="free_y")
```

```{r}
df_variation %>% 
  dplyr::filter(in_cds == TRUE, ref_len_diff > 0) %>% 
  group_by(period) %>% 
  summarize(count = n()) %>%
  ggplot(aes(x="", y=count, fill=as.factor(period))) +
  geom_col(color="black") +
  coord_polar(theta = "y") +
  ggtitle("In exon") + 
  geom_label(aes(label = count),
             color = "white",
             position = position_dodge(width=0.3),
             show.legend = FALSE) +
  theme_void()
```

















