---
title: "STR variation analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load libs}
library(tidyverse)
source("str_gt_utils.R")
```

```{r load variation, echo=FALSE}
# patients with both PT and BDN sample available (10.05.2022)
# shared_patients = c('TCGA-AA-A01K', 'TCGA-AA-3851', 'TCGA-AA-A00O', 'TCGA-AA-3867', 'TCGA-AA-3855', 'TCGA-AA-3860', 'TCGA-AA-3856', 'TCGA-AA-3972', 'TCGA-AA-3949')
# shared_patients =c('TCGA-AA-A02E', 'TCGA-A6-5661', 'TCGA-AA-3968', 'TCGA-AD-A5EK', 'TCGA-AA-3972', 'TCGA-AG-3728', 'TCGA-AA-3855', 'TCGA-AA-3851', 'TCGA-AA-3529', 'TCGA-A6-6780', 'TCGA-AD-6895', 'TCGA-AA-3854', 'TCGA-A6-2675', 'TCGA-A6-5666', 'TCGA-AA-A00O', 'TCGA-A6-3810', 'TCGA-A6-2685', 'TCGA-AA-3517', 'TCGA-AG-3890', 'TCGA-A6-2677', 'TCGA-AF-2690', 'TCGA-A6-A56B', 'TCGA-A6-3807', 'TCGA-A6-2682', 'TCGA-AA-3846', 'TCGA-AA-3522', 'TCGA-AG-3574', 'TCGA-AA-3848', 'TCGA-A6-5664', 'TCGA-A6-3808', 'TCGA-AA-3856', 'TCGA-A6-6653', 'TCGA-A6-5662', 'TCGA-AA-3867', 'TCGA-A6-2674', 'TCGA-AA-3875', 'TCGA-AA-3966', 'TCGA-AA-3524', 'TCGA-AA-A02R', 'TCGA-AA-3531', 'TCGA-A6-6781', 'TCGA-AA-3950', 'TCGA-AA-3514', 'TCGA-AA-A01K', 'TCGA-AA-3527', 'TCGA-A6-5667', 'TCGA-A6-4107', 'TCGA-A6-6652', 'TCGA-A6-A565', 'TCGA-AA-3519', 'TCGA-AA-3530', 'TCGA-AD-6890', 'TCGA-AA-A02W', 'TCGA-A6-5665', 'TCGA-A6-5660', 'TCGA-A6-2681', 'TCGA-AA-A02H', 'TCGA-A6-2680', 'TCGA-A6-6650', 'TCGA-A6-6654', 'TCGA-A6-6651', 'TCGA-A6-2683', 'TCGA-A6-A566', 'TCGA-AA-A02J', 'TCGA-AA-3860', 'TCGA-AA-3949', 'TCGA-AA-A017', 'TCGA-A6-5659', 'TCGA-AA-3866', 'TCGA-A6-6782')

# patients with both PT and BDN sample available (23.05.2022)
shared_patients =c('TCGA-AA-A02E', 'TCGA-A6-6650', 'TCGA-AA-3848', 'TCGA-A6-5665', 'TCGA-A6-6654', 'TCGA-A6-5664', 'TCGA-AA-3684', 'TCGA-AA-3966', 'TCGA-AA-A02H', 'TCGA-AA-3819', 'TCGA-AA-3522', 'TCGA-AA-3680', 'TCGA-A6-6782', 'TCGA-AA-3556', 'TCGA-AA-3968', 'TCGA-AA-3815', 'TCGA-A6-2675', 'TCGA-A6-2677', 'TCGA-AA-3561', 'TCGA-AA-3517', 'TCGA-AA-3693', 'TCGA-A6-6652', 'TCGA-A6-A565', 'TCGA-A6-6651', 'TCGA-AA-3560', 'TCGA-A6-5659', 'TCGA-AA-A01K', 'TCGA-A6-4107', 'TCGA-A6-3810', 'TCGA-AA-3679', 'TCGA-AA-3867', 'TCGA-AA-3538', 'TCGA-AA-A017', 'TCGA-A6-5660', 'TCGA-A6-2685', 'TCGA-A6-3807', 'TCGA-AA-3851', 'TCGA-A6-A56B', 'TCGA-AA-3529', 'TCGA-AA-3818', 'TCGA-AA-A00O', 'TCGA-A6-A566', 'TCGA-AA-A02R', 'TCGA-A6-5662', 'TCGA-AA-3530', 'TCGA-A6-6653', 'TCGA-A6-2674', 'TCGA-AA-3673', 'TCGA-A6-2682', 'TCGA-AG-3574', 'TCGA-A6-5666', 'TCGA-AA-3664', 'TCGA-AA-3949', 'TCGA-AF-2690', 'TCGA-AA-3866', 'TCGA-AA-A02W', 'TCGA-A6-2681', 'TCGA-AD-6890', 'TCGA-AA-3534', 'TCGA-A6-5667', 'TCGA-AA-3821', 'TCGA-AG-3728', 'TCGA-A6-2683', 'TCGA-A6-6780', 'TCGA-AA-3950', 'TCGA-AA-3519', 'TCGA-AA-3814', 'TCGA-A6-2680', 'TCGA-AA-3531', 'TCGA-A6-5661', 'TCGA-A6-6781', 'TCGA-AA-3710', 'TCGA-AA-3852', 'TCGA-AA-3667', 'TCGA-AD-6895', 'TCGA-AA-3527', 'TCGA-AG-3890', 'TCGA-AA-3854', 'TCGA-AA-3514', 'TCGA-AA-3678', 'TCGA-AA-3672', 'TCGA-AA-3846', 'TCGA-A6-3808', 'TCGA-AA-3860', 'TCGA-AA-A02J', 'TCGA-AA-3688', 'TCGA-AA-3855', 'TCGA-AA-3812', 'TCGA-AD-A5EK', 'TCGA-AA-3524', 'TCGA-AA-3696', 'TCGA-AA-3532', 'TCGA-AA-3972', 'TCGA-AA-3856', 'TCGA-AA-3875')

df_genotypes_pt <- read.csv("/Users/maxverbiest/PhD/projects/genotype_strs/results/oxana_variation/crc_primarytumor_19may.csv") %>% 
  filter(patient %in% shared_patients)
df_genotypes_bdn <- read.csv("/Users/maxverbiest/PhD/projects/genotype_strs/results/oxana_variation/crc_bloodnormal_18may.csv") %>% 
  filter(patient %in% shared_patients)

# load variation data
df_genotypes <- read.csv("/Users/maxverbiest/PhD/projects/genotype_strs/results/oxana_variation/crc_solid_april.csv")
df_genotypes <- rbind(df_genotypes, df_genotypes_bdn, df_genotypes_pt)

df_genotypes <- df_genotypes %>% 
  distinct() %>%
  mutate(tmp_id = paste0(chr, "_", start), ref_len_diff = alt - ref)
```


```{r}
# pivot into wider table where bi-allelic tumour and healthy gt are one the same row
df_patient_variation <- df_genotypes %>% 
  mutate(sample_type = if_else(sample_type == "Primary Tumor", "tumor", "healthy")) %>% 
  group_by(patient, sample_type, tmp_id) %>%
  mutate(allele = LETTERS[1:n()]) %>% 
  pivot_wider(
    id_cols = c(patient, sample_type, tmp_id),
    names_from = allele,
    values_from = alt,
    names_prefix = "allele_"
  ) %>% 
  mutate(allele_B = if_else(is.na(allele_B), allele_A, allele_B)) %>% 
  group_by(patient, tmp_id) %>% 
  pivot_wider(
    id_cols = c(patient, tmp_id),
    names_from = sample_type,
    values_from = c(allele_A, allele_B)
  ) %>% ungroup()
```


```{r}
# This filter takes DF from 4222069 -> 2615301 rows
df_patient_variation <- df_patient_variation %>% 
  dplyr::filter(!is.na(allele_A_tumor),!is.na(allele_A_healthy)) %>% 
  dplyr::relocate(allele_A_healthy, .before = allele_B_healthy)

df_patient_variation <- df_patient_variation %>% 
  left_join(df_genotypes %>% select(tmp_id, period, ref, repeat_id) %>% distinct(), by="tmp_id") %>% 
  rowwise() %>% 
  mutate(
    patient_len_diff = calc_patient_len_dif(allele_A_healthy, allele_B_healthy, allele_A_tumor, allele_B_tumor)
    )

write.table(df_patient_variation, "/Users/maxverbiest/PhD/projects/genotype_strs/results/oxana_variation/patient_variation_wide.tsv", quote=F, row.names=F, sep="\t")
```


```{r}
df_patient_variation <- read.csv("/Users/maxverbiest/PhD/projects/genotype_strs/results/oxana_variation/patient_variation_wide.tsv", header=T, sep="\t")
df_str_info <- read.csv("/Users/maxverbiest/PhD/data/str_panels/tral_and_perf_panel_meta_info.tsv", header=T, sep="\t")

# This filter takes DF from 3,767,565 -> 3,559,889 rows
df_patient_variation_filt <- df_patient_variation %>% 
  left_join(df_str_info %>% select(tmp_id, in_segdup), by="tmp_id") %>% 
  filter(!in_segdup) %>% 
  select(-in_segdup)

# This filter takes DF from  3,559,889 -> 2,158,483 rows (remove homorepeats and repeats with close neighbours)
df_patient_variation_filt <- df_patient_variation_filt %>% 
  left_join(df_str_info %>% select(tmp_id, neighbour_type), by="tmp_id") %>% 
  filter(period > 1, neighbour_type == "no_neighbour") %>% 
  select(-neighbour_type)

df_clinical <- read.csv("/Users/maxverbiest/PhD/projects/genotype_strs/data/clinical/oxana_tcga_clinical.csv", header = TRUE) %>%
  dplyr::select(c(case_submitter_id, gender, MSI, CMS))
colnames(df_clinical)[1] <- "patient"
```


```{r}
df_patient_variation_filt %>% 
  left_join(df_clinical, by="patient") %>% 
  group_by(patient) %>% 
  summarize(total_calls = n()) %>% 
  left_join(df_clinical, by="patient") %>% 
  ggplot(aes(x=reorder(patient,total_calls), y=total_calls, fill=MSI)) +
  geom_bar(stat='identity') +
  theme(axis.text.x = element_text(angle=45, hjust=1))

# This removes 19 MSS patients that have very few called loci (84 -> 69 patients)
df_patient_variation_filt <- df_patient_variation_filt %>% 
  group_by(patient) %>% 
  filter(n() > 5000)
```

```{r}
df_patient_variation_filt %>% 
  left_join(df_clinical, by="patient") %>% 
  mutate(variable = ifelse(patient_len_diff > 0, "variable", "not_variable")) %>% 
  group_by(MSI, patient, variable) %>% 
  summarize(count = n()) %>% 
  pivot_wider(
    names_from = variable,
    values_from = count
  ) %>% 
  mutate(perc_variable = variable / (variable + not_variable) * 100) %>%
  ggplot(aes(x=MSI, y=perc_variable, fill=MSI)) +
  geom_boxplot()

df_patient_variation_filt %>% 
  left_join(df_clinical, by="patient") %>% 
  mutate(variable = ifelse(patient_len_diff > 0, "variable", "not_variable")) %>% 
  group_by(MSI, patient, variable) %>% 
  summarize(count = n()) %>% 
  pivot_wider(
    names_from = variable,
    values_from = count
  ) %>% 
  mutate(perc_variable = variable / (variable + not_variable) * 100) %>% 
  ggplot(aes(x=reorder(patient,perc_variable), y=perc_variable, fill=MSI)) +
  geom_bar(stat="identity") +
  theme(axis.text.x = element_text(angle=45, hjust=1))

df_patient_variation_filt %>% 
  left_join(df_clinical, by="patient") %>% 
  mutate(variable = ifelse(patient_len_diff > 0, "variable", "not_variable")) %>% 
  group_by(CMS, patient, variable) %>% 
  summarize(count = n()) %>% 
  pivot_wider(
    names_from = variable,
    values_from = count
  ) %>% 
  mutate(perc_variable = variable / (variable + not_variable) * 100) %>% 
  ggplot(aes(x=reorder(patient,perc_variable), y=perc_variable, fill=CMS)) +
  geom_bar(stat="identity") +
  theme(axis.text.x = element_text(angle=45, hjust=1))
```


```{r}
df_patient_variation_filt %>% 
  left_join(df_clinical, by="patient") %>% 
  group_by(MSI, tmp_id) %>% 
  summarize(count = n()) %>%
  ggplot(aes(x=count)) +
  geom_histogram() +
  facet_wrap(~MSI)

df_patient_variation_filt %>% 
  group_by(tmp_id) %>% 
  summarize(count = n()) %>%
  filter(count >= 10) %>% 
  count()
```


```{r}
df_locus_variability <- df_patient_variation_filt %>%
  left_join(df_clinical, by="patient") %>% 
  mutate(instable = patient_len_diff > 0) %>% 
  group_by(MSI, tmp_id, instable) %>%
  summarize(count=n()) %>% 
  pivot_wider(
    names_from = instable,
    values_from = count,
    names_prefix = "instable_"
  ) %>%
  mutate(
    instable_calls = ifelse(is.na(instable_TRUE), 0, instable_TRUE),
    stable_calls = ifelse(is.na(instable_FALSE), 0, instable_FALSE),
    total_calls = stable_calls + instable_calls,
    frac_variable = instable_calls / total_calls
    ) %>% select(-c(instable_TRUE, instable_FALSE)) %>% ungroup()

df_locus_variability <- df_locus_variability %>% 
  left_join(
    df_patient_variation_filt %>% 
      left_join(
        df_clinical, by="patient"
        ) %>% 
      mutate(instable = patient_len_diff > 0) %>% 
      group_by(MSI, tmp_id) %>% 
      summarize(avg_size_diff = mean(patient_len_diff)), by=c("MSI", "tmp_id")
    )
  
df_locus_variability <- df_locus_variability %>% 
  left_join(df_patient_variation_filt %>% ungroup() %>% select(tmp_id, repeat_id) %>% distinct(), by="tmp_id")

write.table(df_locus_variability, "/Users/maxverbiest/PhD/projects/genotype_strs/results/oxana_variation/locus_variation.tsv", quote=F, row.names=F, sep="\t")
```


```{r}
df_summary  <- data.frame(
  genomic_region = c("all", "UTR", "CDS", "intron/intergenic"),
  total_calls = c(
    df_patient_len_diff_filt %>% count() %>% as.integer(),
    df_patient_len_diff_filt %>% filter(in_exon, !in_cds) %>% count() %>% as.integer(),
    df_patient_len_diff_filt %>% filter(in_cds) %>% count() %>% as.integer(),
    df_patient_len_diff_filt %>% filter(!in_exon) %>% count() %>% as.integer()
  ),
  variable_calls = c(
    df_mutated_strs %>% count() %>% as.integer(),
    df_mutated_strs %>% filter(in_exon, !in_cds) %>% count() %>% as.integer(),
    df_mutated_strs %>% filter(in_cds) %>% count() %>% as.integer(),
    df_mutated_strs %>% filter(!in_exon) %>% count() %>% as.integer()
  )
)

df_summary %>% 
  mutate(perc_variable = variable_calls / total_calls * 100)
```


```{r}
df_patient_variation_filt <- df_patient_variation_filt %>% 
  left_join(df_str_info %>% select(tmp_id, region_type), by="tmp_id")
df_patient_variation_filt$region_type <- factor(as.factor(df_patient_variation_filt$region_type), levels=c("intron/intergenic", "UTR", "CDS"), ordered=T)

df_patient_variation_filt %>% 
  # filter(patient_len_diff > 0) %>% 
  # group_by(gen_region, period) %>% 
  left_join(df_str_info %>% select(tmp_id, region_type), by="tmp_id") %>%
  group_by(region_type, period, patient_len_diff > 0) %>%
  count() %>% 
  left_join(
    df_patient_variation_filt %>% left_join(df_str_info %>% select(tmp_id, region_type), by="tmp_id") %>% group_by(region_type, period) %>% count(),
    # df_patient_variation_filt %>% filter(patient_len_diff > 0) %>% group_by(region_type, period) %>% count(), 
    by=c("region_type", "period")
    ) %>% mutate(perc = (n.x/n.y) * 100) %>% 
  ggplot(aes(x=as.factor(period), y=perc, fill = `patient_len_diff > 0`)) +
  geom_bar(stat='identity') +
  facet_wrap(~region_type)

df_patient_variation_filt %>% 
  filter(patient_len_diff > 0) %>% 
  group_by(region_type, period) %>% 
  count() %>% 
  left_join(
    df_patient_variation_filt %>% group_by(region_type, period) %>% count(),
    by=c("region_type", "period")
    ) %>% mutate(perc_variable = (n.x/n.y) * 100) %>%
  ggplot(aes(x=as.factor(period), y=perc_variable, fill=region_type)) +
  geom_bar(stat='identity', position = position_dodge()) +
  scale_fill_hue()

df_patient_variation_filt %>% 
  left_join(df_clinical, by="patient") %>%
  filter(patient_len_diff > 0) %>% 
  group_by(MSI, region_type, period) %>% 
  count() %>% 
  left_join(
    df_patient_variation_filt %>% 
      left_join(df_clinical, by="patient") %>%
      group_by(MSI, region_type, period) %>% 
      count(),
    by=c("MSI", "region_type", "period")
    ) %>% mutate(perc_variable = (n.x/n.y) * 100) %>%
  ggplot(aes(x=as.factor(period), y=perc_variable, fill=region_type)) +
  geom_bar(stat='identity', position = position_dodge()) +
  scale_fill_hue() +
  facet_wrap(~MSI)
  
df_patient_len_diff_filt %>% 
  left_join(df_clinical, by="patient") %>%
  filter(patient_len_diff > 0, !has_flank_str, period>1) %>% 
  group_by(MSI, gen_region, period) %>% 
  count() %>% 
  left_join(
    df_patient_len_diff_filt %>% 
      filter(!has_flank_str) %>%
      left_join(df_clinical, by="patient") %>% 
      group_by(MSI, gen_region, period) %>% count(),
    by=c("MSI", "gen_region", "period")
    ) %>% mutate(perc_variable = (n.x/n.y) * 100) %>% 
  ggplot(aes(x=as.factor(period), y=perc_variable, fill=gen_region)) +
  geom_bar(stat='identity', position = position_dodge()) +
  scale_fill_hue() +
  facet_wrap(~MSI)
```


```{r}
df_locus_variability %>% 
  filter(total_calls > 5) %>% 
  left_join(df_str_info %>% select(tmp_id, period, ref, region_type), by="tmp_id") %>% 
  group_by(MSI, region_type, period) %>% 
  summarize(perc_variable = mean(frac_variable) * 100) %>% 
  ggplot(aes(x=period, y=perc_variable, fill=region_type)) +
  geom_bar(stat="identity", position = position_dodge()) +
  facet_wrap(~MSI)
```


```{r}
# STRs with flank STRs are more likely to be variable
table_fisher <- df_patient_len_diff_filt %>% 
    group_by(has_flank_str, patient_len_diff > 0) %>%
    # group_by(has_flank_str, period, patient_len_diff > 0) %>%
    count() %>%
    pivot_wider(
      names_from = `patient_len_diff > 0`,
      values_from = n,
      names_prefix = "variable_"
    ) %>% ungroup()

table_fisher %>% 
  mutate(perc_variable = `variable_TRUE` / (`variable_TRUE` + `variable_FALSE`) * 100)

fisher.test(
  table_fisher %>% 
    select(-has_flank_str) %>%
    as.matrix()
)

table_plot <- df_patient_len_diff_filt %>% 
    group_by(has_flank_str, period, patient_len_diff > 0) %>%
    count() %>%
    pivot_wider(
      names_from = `patient_len_diff > 0`,
      values_from = n,
      names_prefix = "variable_"
    ) %>% ungroup() %>% 
    mutate(perc_variable = `variable_TRUE` / (`variable_TRUE` + `variable_FALSE`) * 100)

table_plot %>%
  ggplot(aes(x=as.factor(period), y=perc_variable, fill=has_flank_str)) +
  geom_bar(stat='identity', position=position_dodge()) +
  geom_text(aes(label = paste0(`variable_TRUE`, "/", `variable_TRUE` + `variable_FALSE`)), position=position_dodge(width=0.5), hjust=-0.1) +
  coord_flip() +
  xlab("Unit size") + 
  ylab("Percentage of STRs variable") +
  ylim(0, 45) +
  labs(fill="Has flank STR")
  
# Variations in STRs with flank STRs are significantly longer
t.test(
  x=df_mutated_strs %>% filter(has_flank_str) %>% select(patient_len_diff) %>% unlist() %>%  as.integer(),
  y=df_mutated_strs %>% filter(!has_flank_str) %>% select(patient_len_diff) %>% unlist() %>% as.integer()
  )

df_mutated_strs %>% 
  # filter(patient_len_diff < 100) %>% 
  ggplot(aes(x=has_flank_str, y=log(patient_len_diff))) +
  # ggplot(aes(x=period, y=patient_len_diff)) +
  geom_boxplot() 
```

```{r}
df_patient_variation_filt %>% 
  left_join(df_clinical, by="patient") %>% 
  left_join(df_str_info %>% select(tmp_id, max_p_stretch), by="tmp_id") %>% 
  mutate(variable = ifelse(patient_len_diff > 0, "variable", "not_variable")) %>% 
  group_by(MSI, period, max_p_stretch, variable) %>% 
  summarize(count = n()) %>% 
  pivot_wider(
    names_from = variable,
    values_from = count
  ) %>% 
  mutate(perc_variable = variable / (variable + not_variable) * 100) %>%
  ggplot(aes(x=max_p_stretch, y=perc_variable, colour=MSI)) +
  geom_point() +
  geom_smooth(se=T) +
  facet_wrap(~period, scales="free_x") +
  ylim(0, 100)

df_patient_len_diff_filt %>% 
  left_join(df_clinical, by="patient") %>% 
  mutate(variable = ifelse(patient_len_diff > 0, "variable", "not_variable")) %>% 
  group_by(MSI, period, ref, variable) %>% 
  summarize(count = n()) %>% 
  pivot_wider(
    names_from = variable,
    values_from = count
  ) %>% 
  mutate(perc_variable = variable / (variable + not_variable) * 100) %>% 
  ggplot(aes(x=ref, y=perc_variable, colour=MSI)) +
  geom_point() +
  geom_smooth(se=F) +
  facet_wrap(~period, scales="free_x") +
  ylim(0, 100)

df_patient_variation_filt %>% 
  left_join(df_clinical, by="patient") %>% 
  rowwise() %>% 
  mutate(variable = ifelse(patient_len_diff > 0, "variable", "not_variable"), max_allele = max(allele_A_healthy, allele_B_healthy)) %>% 
  group_by(MSI, period, max_allele, variable) %>% 
  summarize(count = n()) %>% 
  pivot_wider(
    names_from = variable,
    values_from = count
  ) %>% 
  mutate(perc_variable = variable / (variable + not_variable) * 100) %>% 
  ggplot(aes(x=max_allele, y=perc_variable, colour=MSI)) +
  geom_point() +
  geom_smooth(se=F) +
  facet_wrap(~period, scales="free_x") +
  ylim(0, 100)
```













